sdks/go/pkg/beam/core/runtime/harness/hooks.go

```diff
@@ -0,0 +1,151 @@
+// Hooks allow runners to tailor execution of the worker to allow for customization
+// of features used by the harness.
+//
+// Examples of customization:
+//
+// gRPC integration
+// session recording
+// profile recording
+//
+// Registration methods for hooks must be called prior to calling beam.Init()
+// Request methods for hooks must be called as part of building the pipeline
+// request for the runner's Execute method.
+
+package harness
+
+import (
+	"context"
+	"fmt"
+	"io"
+	"time"
+
+	"github.com/apache/beam/sdks/go/pkg/beam/core/runtime"
+	"github.com/apache/beam/sdks/go/pkg/beam/util/grpcx"
+
+	"google.golang.org/grpc"
+)
+
+const hooksNamespaceKey = "__hooks:"
+const sessionCaptureHookKey = hooksNamespaceKey + "session_capture"
+const traceCaptureHookKey = hooksNamespaceKey + "trace_capture"
+const grpcHookKey = hooksNamespaceKey + "grpc"
+
+func bindHooks() {
+	if sessionCapture := runtime.GlobalOptions.Get(sessionCaptureHookKey); sessionCapture != "" {
+		capture = sessionCaptureHook(sessionCapture)
+	}
+
+	if traceCapture := runtime.GlobalOptions.Get(traceCaptureHookKey); traceCapture != "" {
+		profileWriter = traceCaptureHook(traceCapture)
+	}
+
+	if grpc := runtime.GlobalOptions.Get(grpcHookKey); grpc != "" {
+		grpcHooks := grpcHook(grpc)
+		if grpcHooks.Dialer != nil {
+			grpcx.Dial = grpcHooks.Dialer
+		}
+	}
+}
+
+// SessionCaptureHook writes the messaging content consumed and
+// produced by the worker, allowing the data to be used as
+// an input for the session runner. Since workers can exist
+// in a variety of environments, this allows the runner
+// to tailor the behavior best for its particular needs.
+type SessionCaptureHook io.WriteCloser
+
+var sessionCaptureHookRegistry = make(map[string]SessionCaptureHook)
+
+// RegisterSessionCaptureHook registers a SessionCaptureHook for the
+// supplied identifier. It panics if the same identifier is
+// registered twice.
+func RegisterSessionCaptureHook(name string, c SessionCaptureHook) {
```

Comment:
I would expect the structure to be more as follows:
  (1) One can register one or more kinds of hooks. Each kind of hook would implement an interface that is used in  the harness to invoke it. The interface(s) would be general purpose instead of modeling a particular use case. Hook names would be an open scheme.
  (2) Each use case (profiling, session, grpc, etc) can be moved into a separate package each and become fully optional.

For example, a simple kind of hook we might want could be:

type InitHook interface {
      Init(context.Context) error
}

The gRPC implementation would use this hook to set the global variable and the harness would not have to handle that assignment. Ideally, the harness becomes much simpler this way: it just calls a list of hook methods at the appropriate times for each kind.